### PR TITLE
Conditional expression does no longer evaluate both sides.

### LINKED ISSUE #12521: Ternary operation being evaluated when it should not
If you write a ternary in a Linq operation, all the operands will be evaluated by the expression visitor, even though they should not. If any operand happens to be an object property, EF will try to access it even if the ternary specifies not to. If that object is null, the query will throw an exception.

Exception message:
 `An exception was thrown while attempting to evaluate the LINQ query parameter expression 'value(Repro.Program+<>c__DisplayClass0_0).foo.Bar'.`

Stack trace:
```
   at Microsoft.EntityFrameworkCore.Query.ExpressionVisitors.Internal.ParameterExtractingExpressionVisitor.Evaluate(Expression expression, String& parameterName)
   at Microsoft.EntityFrameworkCore.Query.ExpressionVisitors.Internal.ParameterExtractingExpressionVisitor.TryExtractParameter(Expression expression)
   at Microsoft.EntityFrameworkCore.Query.ExpressionVisitors.Internal.ParameterExtractingExpressionVisitor.VisitMember(MemberExpression memberExpression)
   at System.Linq.Expressions.MemberExpression.Accept(ExpressionVisitor visitor)
   at System.Linq.Expressions.ExpressionVisitor.Visit(Expression node)
   at System.Linq.Expressions.ExpressionVisitor.VisitConditional(ConditionalExpression node)
   at System.Linq.Expressions.ConditionalExpression.Accept(ExpressionVisitor visitor)
   at System.Linq.Expressions.ExpressionVisitor.Visit(Expression node)
   at System.Linq.Expressions.ExpressionVisitor.Visit(ReadOnlyCollection`1 nodes)
   at Microsoft.EntityFrameworkCore.Query.ExpressionVisitors.Internal.ParameterExtractingExpressionVisitor.VisitNew(NewExpression node)
   at System.Linq.Expressions.NewExpression.Accept(ExpressionVisitor visitor)
   at System.Linq.Expressions.ExpressionVisitor.Visit(Expression node)
   at System.Linq.Expressions.ExpressionVisitor.VisitLambda[T](Expression`1 node)
   at Microsoft.EntityFrameworkCore.Query.ExpressionVisitors.Internal.ParameterExtractingExpressionVisitor.VisitLambda[T](Expression`1 node)
   at System.Linq.Expressions.Expression`1.Accept(ExpressionVisitor visitor)
   at System.Linq.Expressions.ExpressionVisitor.Visit(Expression node)
   at System.Linq.Expressions.ExpressionVisitor.VisitUnary(UnaryExpression node)
   at Microsoft.EntityFrameworkCore.Query.ExpressionVisitors.Internal.ParameterExtractingExpressionVisitor.VisitUnary(UnaryExpression unaryExpression)
   at System.Linq.Expressions.UnaryExpression.Accept(ExpressionVisitor visitor)
   at System.Linq.Expressions.ExpressionVisitor.Visit(Expression node)
   at System.Dynamic.Utils.ExpressionVisitorUtils.VisitArguments(ExpressionVisitor visitor, IArgumentProvider nodes)
   at System.Linq.Expressions.ExpressionVisitor.VisitMethodCall(MethodCallExpression node)
   at Microsoft.EntityFrameworkCore.Query.ExpressionVisitors.Internal.ParameterExtractingExpressionVisitor.VisitMethodCall(MethodCallExpression methodCallExpression)
   at System.Linq.Expressions.MethodCallExpression.Accept(ExpressionVisitor visitor)
   at System.Linq.Expressions.ExpressionVisitor.Visit(Expression node)
   at System.Dynamic.Utils.ExpressionVisitorUtils.VisitArguments(ExpressionVisitor visitor, IArgumentProvider nodes)
   at System.Linq.Expressions.ExpressionVisitor.VisitMethodCall(MethodCallExpression node)
   at Microsoft.EntityFrameworkCore.Query.ExpressionVisitors.Internal.ParameterExtractingExpressionVisitor.VisitMethodCall(MethodCallExpression methodCallExpression)
   at System.Linq.Expressions.MethodCallExpression.Accept(ExpressionVisitor visitor)
   at System.Linq.Expressions.ExpressionVisitor.Visit(Expression node)
   at Microsoft.EntityFrameworkCore.Query.ExpressionVisitors.Internal.ParameterExtractingExpressionVisitor.ExtractParameters(Expression expression)
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryModelGenerator.ExtractParameters(IDiagnosticsLogger`1 logger, Expression query, IParameterValues parameterValues, Boolean parameterize, Boolean generateContextAccessors)
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.Execute[TResult](Expression query)
   at Microsoft.EntityFrameworkCore.Query.Internal.EntityQueryProvider.Execute[TResult](Expression expression)
   at System.Linq.Queryable.FirstOrDefault[TSource](IQueryable`1 source, Expression`1 predicate)
   at _12522Repro.Program.Main(String[] args) in C:\Users\fschlaef\Documents\Visual Studio 2017\Projects\Repro\Repro\Program.cs:line 33
```

### Steps to reproduce
Quick example : `someBoolean ? foo.Bar : 1`

If `someBoolean` is `false` and `foo` is `null`, the query will crash.

Full repro :

```c#
using Microsoft.EntityFrameworkCore;
using Microsoft.Extensions.Logging;
using System;
using System.Diagnostics;
using System.Linq;

namespace _12521Repro
{
    class Foo
    {
        public int Bar { get; set; }
    }

    class Program
    {
        static void Main(string[] args)
        {
            using (var context = new MyContext())
            {
                context.Database.EnsureDeleted();
                context.Database.EnsureCreated();
                context.Customer.Add(new Customer());
                context.SaveChanges();
            }

            try
            {
                using (var context = new MyContext())
                {
                    Foo foo = null;
                    var hasData = !(foo is null);
                    int control = (hasData ? foo.Bar : 1); // Equals 1, our ternary is correct

                    var query = context.Customer
                        .Select(c => new
                        {
                            c.Id,
                            Data1 = hasData ? foo.Bar : 1, // Will crash
                            Data2 = foo != null ? foo.Bar : 1, // Will also crash
                            Data3 = !hasData ? 1 : foo.Bar // Order doesn't matter, crashes
                        })
                        .FirstOrDefault(c => c.Id == 1);
                }
            }
            catch(Exception e)
            {
                Debug.WriteLine(e.Message);
            }
        }
    }

    public class MyContext : DbContext
    {
        public virtual DbSet<Customer> Customer { get; set; }

        protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
        {
            string connectionString = "Server=.;Database=Repro12521;Trusted_Connection=True;MultipleActiveResultSets=true";

            optionsBuilder.UseSqlServer(connectionString)
                .EnableSensitiveDataLogging()
                .UseLoggerFactory(new LoggerFactory().AddDebug(LogLevel.Debug));
        }
    }


    public partial class Customer
    {
        public int Id { get; set; }
    }
}
```
### Workaround

The easiest workaround I could find was to reinstanciate the object manually before the query :

```csharp
if(!hasData)
{
    foo = new Foo();
}
```

This works but this is a very simple case, the solution might not be applicable for more complex cases.

### Further technical details
EF Core version: 2.1.1
Database Provider:  Microsoft.EntityFrameworkCore.SqlServer
Operating system: Windows 7 x64
IDE: Visual Studio 2017 15.7.4

### PR DESCRIPTION
**Pull request by Matthiee**
Conditional expression will no longer evaluate both sides of the expression. 

Fixes #12521

### CODE REVIEW

**Comment by Matthiee** [comment]
Some tests are failing, but I _think_ they are improvements. @smitpatel  please advice on how to procede. 

So what is happening the following code used to produce a different query

### Code
```
var flag = true;

return AssertQuery<Product>(
     isAsync,
     ps => ps
          .Where(p => flag ? p.UnitsInStock >= 20 : p.UnitsInStock < 20),
    entryCount: 51);
```

#### Old query
```
@__flag_0='True'

SELECT [p].[ProductID], [p].[Discontinued], [p].[ProductName], [p].[SupplierID], [p].[UnitPrice], [p].[UnitsInStock]
FROM [Products] AS [p]
WHERE ((@__flag_0 = 1) AND ([p].[UnitsInStock] >= CAST(20 AS smallint))) OR ((@__flag_0 <> 1) AND ([p].[UnitsInStock] < CAST(20 AS smallint)))
```

#### New query
```
SELECT [p].[ProductID], [p].[Discontinued], [p].[ProductName], [p].[SupplierID], [p].[UnitPrice], [p].[UnitsInStock]
FROM [Products] AS [p]
WHERE [p].[UnitsInStock] >= CAST(20 AS smallint)
```

**Comment by smitpatel** [comment]
@Matthiee - Change flag to `p.Discountinued` in that test. Fix is improving the existing test. Modification will make sure it covers old cold path which it is verifying.

**Comment by smitpatel** [comment]
@Matthiee - You can remove one of the tests. And convert flag to `p.Discontinued`
The test covers code path, `a ? b : c` translates to `a && b OR !a && c`

**Comment by Matthiee** [comment]
@smitpatel  - I'm going through some of the failed tests and I just want to be sure if the result is expected.

The following test 
```
[Fact]
public virtual void Where_nested_conditional_search_condition_in_result()
{
    var prm1 = true;
    var prm2 = false;
    var list = new[] { "Foo", "Bar" };

    AssertQuery<NullSemanticsEntity1>(
        es => es.Where(
            e => prm1
                ? (prm2
                    ? (e.BoolA
                        ? e.StringA.StartsWith("A")
                        : false)
                    : true)
                : (e.BoolB ? list.Contains(e.StringA) : list.Contains(e.StringB))));
}
```

Used to produce
```
@__prm1_0='True'
@__prm2_1='False'

SELECT [e].[Id]
FROM [Entities1] AS [e]
WHERE CASE
    WHEN @__prm1_0 = 1
    THEN CASE
        WHEN @__prm2_1 = 1
        THEN CASE
            WHEN [e].[BoolA] = 1
            THEN CASE
                WHEN [e].[StringA] LIKE N'A' + N'%' AND (LEFT([e].[StringA], LEN(N'A')) = N'A')
                THEN CAST(1 AS BIT) ELSE CAST(0 AS BIT)
            END ELSE CAST(0 AS BIT)
        END ELSE CAST(1 AS BIT)
    END ELSE CASE
        WHEN [e].[BoolB] = 1
        THEN CASE
            WHEN [e].[StringA] IN (N'Foo', N'Bar')
            THEN CAST(1 AS BIT) ELSE CAST(0 AS BIT)
        END ELSE CASE
            WHEN [e].[StringB] IN (N'Foo', N'Bar')
            THEN CAST(1 AS BIT) ELSE CAST(0 AS BIT)
        END
    END
END = 1
```

And after the changes it produces the following
```
SELECT [e].[Id]
FROM [Entities1] AS [e]
WHERE CASE
    WHEN 0 = 1
    THEN CASE
        WHEN [e].[BoolA] = 1
        THEN CASE
            WHEN [e].[StringA] LIKE N'A' + N'%' AND (LEFT([e].[StringA], LEN(N'A')) = N'A')
            THEN CAST(1 AS BIT) ELSE CAST(0 AS BIT)
        END ELSE CAST(0 AS BIT)
    END ELSE CAST(1 AS BIT)
END = 1
```

**Comment by smitpatel** [comment]
@Matthiee - That SQL should get even improved. Let me show you where is the bug in code fix.

File: src/EFCore/Query/ExpressionVisitors/Internal/ParameterExtractingExpressionVisitor.cs [hunk 1]
```diff
@@ -294,6 +294,22 @@ protected override Expression VisitMember(MemberExpression memberExpression)
...
+        protected override Expression VisitConditional(ConditionalExpression conditionalExpression)
+        {
+            var returnValue = (TryOptimize(conditionalExpression.Test) as ConstantExpression)?.Value as bool?;
+
+            if (returnValue == null)
+            {
+                return base.VisitConditional(conditionalExpression);
+            }
+
+            return (bool)returnValue ? conditionalExpression.IfTrue : conditionalExpression.IfFalse;
```
Comment (smitpatel) [inline src/EFCore/Query/ExpressionVisitors/Internal/ParameterExtractingExpressionVisitor.cs]: Here you haven't visited IfTrue/IfFalse part of the expression.
```suggestion
            return Visit((bool)returnValue ? conditionalExpression.IfTrue : conditionalExpression.IfFalse);
```

**Comment by smitpatel** [comment]
@Matthiee - I am changing base branch to master, we are not making changes to query in release/2.2 unless critical.

File: src/EFCore/Query/ExpressionVisitors/Internal/ParameterExtractingExpressionVisitor.cs [hunk 2]
```diff
@@ -294,6 +294,22 @@ protected override Expression VisitMember(MemberExpression memberExpression)
             return ExtractParameters(queryable.Expression);
         }
 
+        /// <summary>
+        ///     This API supports the Entity Framework Core infrastructure and is not intended to be used
+        ///     directly from your code. This API may change or be removed in future releases.
+        /// </summary>
+        protected override Expression VisitConditional(ConditionalExpression conditionalExpression)
```
Comment (smitpatel) [inline src/EFCore/Query/ExpressionVisitors/Internal/ParameterExtractingExpressionVisitor.cs]: I thought about this more in connection of the test which changed. This is actually incorrect fix.
All we should do here is to check if conditionalExpression is evaluatable then do tryextractparameter. Similar to how ArrayIndex is working and you are applying fix for coalesce.

The reason is, the test whose sql is changing, by effectively trimming tree of range variable, now we have 2 different trees for different values of same parameter, which causes us to recompile the tree again, effectively defeating the purpose of parameterization.
```suggestion
        protected override Expression VisitConditional(ConditionalExpression conditionalExpression)
        {
            if (_partialEvaluationInfo.IsEvaluatableExpression(conditionalExpression)
            {
                return TryExtractParameter(conditionalExpression);
            }
            
            return base.VisitConditional(conditionalExpression);
        }
```

File: src/EFCore.Specification.Tests/Query/SimpleQueryTestBase.Where.cs [hunk 3]
```diff
@@ -1586,71 +1586,40 @@ public virtual Task Where_concat_string_int_comparison3(bool isAsync)
 
         [ConditionalTheory]
         [MemberData(nameof(IsAsyncData))]
-        public virtual Task Where_ternary_boolean_condition_true(bool isAsync)
+        public virtual Task Where_ternary_boolean_condition(bool isAsync)
```
Comment (smitpatel) [inline src/EFCore.Specification.Tests/Query/SimpleQueryTestBase.Where.cs]: You don't need to make change of flag anymore. All such tests should remain as is.

File: test/EFCore.InMemory.FunctionalTests/Query/QueryBugsInMemoryTest.cs [hunk 4]
```diff
@@ -20,6 +20,55 @@ namespace Microsoft.EntityFrameworkCore.Query
 {
     public class QueryBugsInMemoryTest : IClassFixture<InMemoryFixture>
     {
+        #region Bug12521
+
+        [Fact]
+        public void Ternary_operation_should_not_evaluate_both_sides()
```
Comment (smitpatel) [inline test/EFCore.InMemory.FunctionalTests/Query/QueryBugsInMemoryTest.cs]: Can you also move this to `SimpleQueryTestBase`?

**Pull request merged**